src/test/java/org/apache/sling/jcr/repoinit/CreateGroupWithExternalIdTest.java

```diff
@@ -0,0 +1,131 @@
+
+        U.parseAndExecute("create group " + GROUP_NAME);
+        U.assertGroup(GROUP_NAME + " should exist", GROUP_NAME, true);
+
+        assertThrows(ConstraintViolationException.class, () -> U.parseAndExecute(
+                "set properties on authorizable(" + GROUP_NAME + ")",
+                "  set rep:externalId to \"" + GROUP_NAME + ";ldap\"",
+                "end"
+        ));
+        // The ConstraintViolationException does not refresh the session automatically,
+        // so the invalid changes are still transiently present and need to be explicitly dropped.
+        U.getAdminSession().refresh(false);
+
+        final Authorizable group = UserUtil.getAuthorizable(U.getAdminSession(), GROUP_NAME);
+        assertNull(GROUP_NAME + " should not have a rep:externalId property", group.getProperty("rep:externalId"));
+    }
+
+    @Test
+    public void setRepExternalIdPropertyOnNewGroupSucceeds() throws RepositoryException, RepoInitParsingException {
+        TestUtil U = createTestUtil();
+
+        U.assertGroup(GROUP_NAME + " should not yet exist", GROUP_NAME, false);
+
+        // Do this twice in order to test initial creation and update scenarios.
+        for (int i = 0; i < 2; i++) {
+            // WARNING: rep:externalId is an implementation detail of oak-auth-external and should
+            //          normally only be set by that module. It is NOT recommended to use the approach
+            //          illustrated below to set up a production system.
+            U.parseAndExecute(
+                    "create group " + GROUP_NAME,
+                    "set properties on authorizable(" + GROUP_NAME + ")",
+                    "  set rep:externalId to \"" + GROUP_NAME + ";ldap\"",
+                    "end"
+            );
+            U.assertGroup(GROUP_NAME + " should have been created", GROUP_NAME, true);
+
+            final Authorizable group = UserUtil.getAuthorizable(U.getAdminSession(), GROUP_NAME);
+            final Value[] externalIdValues = group.getProperty("rep:externalId");
+            assertEquals(1, externalIdValues.length);
+            assertEquals(GROUP_NAME + ";ldap", externalIdValues[0].getString());
+        }
+    }
+
+    @NotNull
+    private static TestUtil createTestUtil() throws RepositoryException {
+        final SecurityProvider securityProvider = createSecurityProvider();
+
+        final Repository repository = new Jcr()
+                .with(securityProvider)
+                .createRepository();
+        final Session adminSession = repository.login(new SimpleCredentials("admin", "admin".toCharArray()));
+        return new TestUtil(adminSession);
+    }
+
+    @NotNull
+    private static SecurityProvider createSecurityProvider() {
+        final SecurityProvider defaults = SecurityProviderBuilder.newBuilder().build();
+
+
+        final CompositePrincipalConfiguration principalConfiguration = new CompositePrincipalConfiguration();
```

Comment:
hi @jsedding , i might be mistaken but i believe lines 118-129 can be simplified by using the SecurityProviderHelper to inject just the ExternalPrincipalConfiguration in addition to the default.
that utility class is located in the oak-core test bundle. then you don't have to construct a new SecurityProvider again. wdyt?

Reply:
I'll try that!

Reply:
Ok, that was easy :)